rabbit/rabbittransport.go

```diff
@@ -0,0 +1,426 @@
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	log "github.com/cihub/seelog"
+	uuid "github.com/nu7hatch/gouuid"
+	"github.com/streadway/amqp"
+	"gopkg.in/tomb.v2"
+
+	"github.com/mondough/typhon/errors"
+	"github.com/mondough/typhon/message"
+	"github.com/mondough/typhon/transport"
+)
+
+const (
+	connectTimeout  = 30 * time.Second
+	chanSendTimeout = 10 * time.Second
+	respondTimeout  = 10 * time.Second
+)
+
+var (
+	ErrCouldntConnect   = errors.InternalService("Could not connect to RabbitMQ")
+	ErrDeliveriesClosed = errors.InternalService("Delivery channel closed")
+	ErrNoReplyTo        = errors.BadRequest("Request does not have appropriate X-Rabbit-ReplyTo header")
+)
+
+type rabbitTransport struct {
+	tomb          *tomb.Tomb
+	connM         sync.RWMutex                       // protects conn + connReady
+	conn          *RabbitConnection                  // underlying connection
+	connReady     chan struct{}                      // swapped along with conn (reconnecting)
+	replyQueue    string                             // message reply queue name
+	inflightReqs  map[string]chan<- message.Response // correlation id: response chan
+	inflightReqsM sync.Mutex                         // protects inflightReqs
+	listeners     map[string]*tomb.Tomb              // service name: tomb
+	listenersM    sync.RWMutex                       // protects listeners
+}
+
+// run starts the asynchronous run-loop connecting to RabbitMQ
+func (t *rabbitTransport) run() {
+	initConn := func() *RabbitConnection {
+		conn := NewRabbitConnection()
+		t.connM.Lock()
+		defer t.connM.Unlock()
+		t.conn = conn
+		select {
+		case <-t.connReady:
+			// Only swap connReady if it's already closed
+			t.connReady = make(chan struct{})
+		default:
+		}
+		return conn
+	}
+	conn := initConn()
+
+	t.tomb.Go(func() error {
+		defer func() {
+			t.killListeners()
+			conn.Close()
+			log.Info("[Typhon:RabbitTransport] Dead; connection closed")
+		}()
+
+	runLoop:
+		for {
+			log.Info("[Typhon:RabbitTransport] Run loop connecting…")
+			select {
+			case <-t.tomb.Dying():
+				return nil
+
+			case <-conn.Init():
+				log.Info("[Typhon:RabbitTransport] Run loop connected")
+				t.listenReplies()
+
+				select {
+				case <-t.tomb.Dying():
+					// Do not loop again
+					return nil
+				default:
+					conn.Close()
+					conn = initConn()
+					continue runLoop
+				}
+
+			case <-time.After(connectTimeout):
+				log.Criticalf("[Typhon:RabbitTransport] Run loop timed out after %s waiting to connect",
+					connectTimeout.String())
+				return ErrCouldntConnect
+			}
+		}
+	})
+}
+
+// deliveryChan returns the name of a delivery channel for a given service
+func (t *rabbitTransport) deliveryChan(serviceName string) string {
+	return serviceName
+}
+
+func (t *rabbitTransport) Tomb() *tomb.Tomb {
+	return t.tomb
+}
+
+func (t *rabbitTransport) connection() *RabbitConnection {
+	t.connM.RLock()
+	defer t.connM.RUnlock()
+	return t.conn
+}
+
+func (t *rabbitTransport) Ready() <-chan struct{} {
```

Comment:
Might be worth adding a comment here to remind people (ie. me :wink:) that this returns a channel that will be closed, rather than received off. :+1: on design though.